api/versions.go

```diff
@@ -288,6 +288,95 @@ func (api *DatasetAPI) putVersion(w http.ResponseWriter, r *http.Request) {
 	log.InfoCtx(ctx, "putVersion endpoint: request successful", data)
 }
 
+func (api *DatasetAPI) detachVersion(w http.ResponseWriter, r *http.Request) {
+
+	defer request.DrainBody(r)
+
+	ctx := r.Context()
+	vars := mux.Vars(r)
+
+	log.InfoCtx(ctx, "detachVersion endpoint: endpoint called.", nil)
+
+	datasetID := vars["dataset_id"]
+	edition := vars["edition"]
+	version := vars["version"]
+
+	auditParams := common.Params{"dataset_id": datasetID, "edition": edition, "version": version}
+	logData := audit.ToLogData(auditParams)
+
+	if auditErr := api.auditor.Record(ctx, detachVersionAction, audit.Attempted, auditParams); auditErr != nil {
+		handleVersionAPIErr(ctx, auditErr, w, logData)
+		return
+	}
+
+	err := func() error {
+
+		authorised, logData := api.authenticate(r, logData)
+		if !authorised {
+			log.ErrorCtx(ctx, errors.WithMessage(errs.ErrUnauthorised, "detachVersion endpoint: User is not authorised to detach a dataset version"), logData)
+			return errs.ErrNotFound
+		}
+
+		editionDoc, err := api.dataStore.Backend.GetEdition(datasetID, edition, models.PublishedState)
+		if err != nil {
+			log.ErrorCtx(ctx, errors.WithMessage(errs.ErrEditionNotFound, "detachVersion endpoint: Cannot find the specified edition"), logData)
+			return err
+		}
+
+		// Only permit detachment of the latest version.
+		if editionDoc.Next.Links.LatestVersion.ID != version {
+			log.ErrorCtx(ctx, errors.WithMessage(errs.ErrVersionAlreadyExists, "detachVersion endpoint: Detach called againt a version other than latest, aborting."), logData)
+			return errs.ErrVersionAlreadyExists
+		}
+
+		// Only permit detachment where state is edition-confirmed or associated
+		state := editionDoc.Next.State
+		if state != models.AssociatedState && state != models.EditionConfirmedState {
+			log.ErrorCtx(ctx, errors.WithMessage(errs.ErrIncorrectStateToDetach, "detachVersion endpoint: You can only detach a version with a state of edition-confirmed or associated"), logData)
+			return errs.ErrIncorrectStateToDetach
+		}
+
+		versionDoc, err := api.dataStore.Backend.GetVersion(datasetID, edition, version, editionDoc.Next.State)
+		if err != nil {
+			log.ErrorCtx(ctx, errors.WithMessage(errs.ErrVersionNotFound, "detachVersion endpoint: Cannot find the specified edition"), logData)
+			return errs.ErrVersionNotFound
+		}
+
+		// Detach the version
+		versionDoc.State = models.DetachedState
+		if err = api.dataStore.Backend.UpdateVersion(versionDoc.ID, versionDoc); err != nil {
+			log.ErrorCtx(ctx, errors.WithMessage(err, "detachVersion endpoint: failed to update version document"), logData)
+			return err
+		}
+
+		// Rollback the edition
```

Comment:
Do we need to rollback the dataset as well? It will have a latest_version link too

Reply:
fair point. Have added it below.